pkg/controller/scylladbdatacenter/status.go

```diff
@@ -0,0 +1,138 @@
+package scylladbdatacenter
+
+import (
+	"context"
+	"fmt"
+
+	scyllav1alpha1 "github.com/scylladb/scylla-operator/pkg/api/scylla/v1alpha1"
+	"github.com/scylladb/scylla-operator/pkg/naming"
+	"github.com/scylladb/scylla-operator/pkg/pointer"
+	appsv1 "k8s.io/api/apps/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/klog/v2"
+)
+
+func (scc *Controller) updateStatus(ctx context.Context, currentSC *scyllav1alpha1.ScyllaDBDatacenter, status *scyllav1alpha1.ScyllaDBDatacenterStatus) error {
+	if apiequality.Semantic.DeepEqual(&currentSC.Status, status) {
+		return nil
+	}
+
+	sdc := currentSC.DeepCopy()
+	sdc.Status = *status
+
+	// Make sure that any "live" updates to the status are always manifested in the aggregated fields.
+	updateAggregatedStatusFields(&sdc.Status)
+
+	klog.V(2).InfoS("Updating status", "ScyllaDBDatacenter", klog.KObj(sdc))
+
+	_, err := scc.scyllaClient.ScyllaDBDatacenters(sdc.Namespace).UpdateStatus(ctx, sdc, metav1.UpdateOptions{})
+	if err != nil {
+		return err
+	}
+
+	klog.V(2).InfoS("Status updated", "ScyllaDBDatacenter", klog.KObj(sdc))
+
+	return nil
+}
+
+func (scc *Controller) getScyllaVersion(sts *appsv1.StatefulSet) (string, error) {
+	firstMemberName := fmt.Sprintf("%s-0", sts.Name)
+	firstMember, err := scc.podLister.Pods(sts.Namespace).Get(firstMemberName)
+	if err != nil {
+		return "", err
+	}
+
+	controllerRef := metav1.GetControllerOfNoCopy(firstMember)
+	if controllerRef == nil || controllerRef.UID != sts.UID {
+		return "", fmt.Errorf("foreign pod")
+	}
+
+	version, err := naming.ScyllaVersion(firstMember.Spec.Containers)
+	if err != nil {
+		return "", err
+	}
+
+	return version, nil
+}
+
+// calculateRackStatus calculates a status for the rack.
+// sts and old status may be nil.
+func (scc *Controller) calculateRackStatus(sdc *scyllav1alpha1.ScyllaDBDatacenter, sts *appsv1.StatefulSet) *scyllav1alpha1.RackStatus {
+	status := &scyllav1alpha1.RackStatus{
+		Nodes:          pointer.Ptr(int32(0)),
+		CurrentNodes:   pointer.Ptr(int32(0)),
+		UpdatedNodes:   pointer.Ptr(int32(0)),
+		ReadyNodes:     pointer.Ptr(int32(0)),
+		AvailableNodes: pointer.Ptr(int32(0)),
+		Stale:          pointer.Ptr(true),
+	}
+
+	if sts == nil {
+		return status
+	}
+
+	status.Name = sts.Labels[naming.RackNameLabel]
+	status.Nodes = pointer.Ptr(*sts.Spec.Replicas)
+	status.ReadyNodes = pointer.Ptr(sts.Status.ReadyReplicas)
+	status.AvailableNodes = pointer.Ptr(sts.Status.AvailableReplicas)
+	status.UpdatedNodes = pointer.Ptr(sts.Status.UpdatedReplicas)
+	status.CurrentNodes = pointer.Ptr(sts.Status.CurrentReplicas)
+	status.Stale = pointer.Ptr(sts.Status.ObservedGeneration < sts.Generation)
+
+	scyllaDBImageVersion, err := naming.ImageToVersion(sdc.Spec.ScyllaDB.Image)
```

Comment:
this is going to fail for non-tagged image references, or even tagged ones with a digest - wasn't the new api supposed to support any image reference format?

Reply:
This PR doesn't extend supported image types. It prepares API to allow it.

Reply:
It could use at least a todo then

Reply:
It's not the only place requiring changing and which are is not obvious.

Reply:
ok